### PR TITLE
feat(cron):periodically run jobs by expanding cronPolicy in job specification

### LINKED ISSUE #218: [feature request] periodically run jobs by expanding cronPolicy in job specification
**What would you like to be added**:

extend job api and introduce `cronPolicy` to setup cron schedule, trigger running jobs periodically, for example:

```yaml
spec:
  runPolicy:
    cronPolicy:
      schedule: "0 * * * *"
      concurrentPolicy: Allow
  tfReplicaSpecs:
    Worker:
      ...
```

**Why is this needed**:

### PR DESCRIPTION
**Pull request by hoaresky**
### Ⅰ. Describe what this PR does
Periodically run jobs by expanding cronPolicy in job specification

### II. Does this pull request fix one issue?

fix #218

### CODE REVIEW

File: pkg/job_controller/job.go [hunk 1]
```diff
@@ -324,6 +330,108 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
...
+		return err
+	}
+	if !isCronJob {
+		//Job no longer needs to run regularly, delete the originally created cronJob
+		return jc.Client.Delete(context.Background(), cronJob)
+	}
+	//If cronPolicy or cronTemplate change, the cronJob needs to be updated
+	newCronJob := cronJob.DeepCopy()
+	needUpdate := false
+	if !reflect.DeepEqual(cronJob.Spec.CronPolicy, *policy.CronPolicy) {
```
Comment (SimonCqk) [inline pkg/job_controller/job.go:360]: you may compare `newInstance.Spec` with `cronJob.Spec` directly , and two-step `DeepEqual` comparison can be simplified into one.

File: pkg/job_controller/job.go [hunk 2]
```diff
@@ -324,6 +330,108 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
...
+	extCodec := pkgruntime.NewRawExtensionCodec(jc.Scheme)
+	newMeta := runtimeObj.DeepCopyObject().(metav1.Object)
+	metaName := newMeta.GetName()
+	metaNs := newMeta.GetNamespace()
+	// The new Cron only need Spec info, but Spec cannot get here,
+	// it can only be converted into metaObj by force, then cleared
+	// some attributes of meta. The overall framework will be refactored
+	// and optimized in the future
+	clearMetaInfoForCronTemplate(newMeta)
+	rawObj, err := extCodec.EncodeRaw(newMeta.(runtime.Object))
```
Comment (SimonCqk) [inline pkg/job_controller/job.go:398]: extCodec.EncodeRaw(runtimeObject)

File: pkg/job_controller/job.go [hunk 3]
```diff
@@ -324,6 +330,108 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
...
+			CronTemplate: cronTemplateSpec,
+		},
+	}
+	return instance, nil
+}
+
+// Extract the submitted job configuration to generate a cron
+// job configuration, and set the parameters that cannot
+// be assigned to empty.
+func clearMetaInfoForCronTemplate(newMeta metav1.Object) {
```
Comment (SimonCqk) [inline pkg/job_controller/job.go:426]: func resetNonInheritableMetaFields (metav1.Object) ?

File: pkg/job_controller/job.go [hunk 4]
```diff
@@ -324,6 +330,108 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
...
+		return jc.Client.Delete(context.Background(), cronJob)
+	}
+	//If cronPolicy or cronTemplate change, the cronJob needs to be updated
+	newCronJob := cronJob.DeepCopy()
+	needUpdate := false
+	if !reflect.DeepEqual(cronJob.Spec.CronPolicy, *policy.CronPolicy) {
+		newCronJob.Spec.CronPolicy = *policy.CronPolicy
+		needUpdate = true
+	}
+	newInstance, err := jc.createCronInstance(policy, runtimeObj)
```
[the remark this answers is not in the document]
Comment (hoaresky) [reply pkg/job_controller/job.go:364]: It can't be merged because one is used to create and  another one  is used to update

File: pkg/job_controller/job.go (hunk 2 above)
Comment (hoaresky) [reply pkg/job_controller/job.go:398]: Need to encode a cleaned-up runtimeObj, the input runtimeObj contains some parameters that should not be inherited

File: pkg/job_controller/job_controller.go [hunk 5]
```diff
@@ -90,6 +90,9 @@ type JobController struct {
 
 	// Client talks to api-server
 	Client client.Client
+
+	//Scheme
```
Comment (SimonCqk) [inline pkg/job_controller/job_controller.go:94]: Add more comment for `Scheme` field

File: pkg/job_controller/job.go [hunk 6]
```diff
@@ -84,6 +88,10 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
 	}
 	log.Infof("Reconciling for job %s", metaObject.GetName())
 
+	//if it's a scheduled task，create a cronJob of the same name
+	if err := jc.ReconcileCron(metaObject, runtimeObject, runPolicy); err != nil {
```
Comment (jian-he) [inline pkg/job_controller/job.go:92]: 这里加个check, 如果 cron没有define，不用reconcile

File: pkg/job_controller/job.go [hunk 7]
```diff
@@ -324,6 +332,116 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
...
+	if *tempPolicy.CronPolicy.PropagationDelete {
+		jobReference := metav1.OwnerReference{
+			APIVersion:         apiversion,
+			Kind:               kind,
+			Name:               metaName,
+			UID:                metaUid,
+			Controller:         pointer.BoolPtr(true),
+			BlockOwnerDeletion: pointer.BoolPtr(true),
+		}
+		if instance.OwnerReferences != nil {
```
Comment (jian-he) [inline pkg/job_controller/job.go:424]: 这个 instance 是新创建的，需要这些check么？

File: pkg/job_controller/job.go [hunk 8]
```diff
@@ -324,6 +332,116 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      metaName,
+			Namespace: metaNs,
+		},
+		Spec: appv1.CronSpec{
+			CronPolicy:   *tempPolicy.CronPolicy,
+			CronTemplate: cronTemplateSpec,
+		},
+	}
+	if *tempPolicy.CronPolicy.PropagationDelete {
```
Comment (jian-he) [inline pkg/job_controller/job.go:415]: PropagationDelete 是啥意思，把cronInstance的 owner设置为 job ? 这个意思是说job delete, cron 会被delete.

File: pkg/job_controller/job.go [hunk 9]
```diff
@@ -324,6 +332,116 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
...
+		newCronJob.Spec.CronTemplate = newInstance.Spec.CronTemplate
+		return jc.Client.Patch(context.Background(), newCronJob, client.MergeFrom(cronJob))
+	}
+	return nil
+}
+
+func (jc *JobController) createCronInstance(policy *apiv1.RunPolicy, runtimeObj runtime.Object) (*appv1.Cron, error) {
+	tempPolicy := policy.DeepCopy()
+	defer func() {
+		policy.CronPolicy = tempPolicy.CronPolicy
```
Comment (jian-he) [inline pkg/job_controller/job.go:375]: why is this needed ? already made a deepcopy

File: pkg/job_controller/job.go [hunk 10]
```diff
@@ -324,6 +332,116 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
...
+		// Job no longer needs to run regularly, delete the originally created cronJob
+		return jc.Client.Delete(context.Background(), cronJob)
+	}
+	// If cronPolicy or cronTemplate change, the cronJob needs to be updated
+	newCronJob := cronJob.DeepCopy()
+	newInstance, err := jc.createCronInstance(policy, runtimeObj)
+	if err != nil {
+		return err
+	}
+	if !reflect.DeepEqual(cronJob.Spec, newInstance.Spec) {
```
Comment (jian-he) [inline pkg/job_controller/job.go:365]: 这里比较的是spec，下面赋值的是cronTemplate

File: pkg/job_controller/job.go (hunk 6 above)
Comment (hoaresky) [reply pkg/job_controller/job.go:92]: When the job is submitted with cron expression for the first time, a cronJob is generated, then the second modification deletes the cron expression submission. If the check logic is added here, the generated cron Job cannot be deleted,not as expected

File: pkg/job_controller/job.go (hunk 8 above)
Comment (hoaresky) [reply pkg/job_controller/job.go:415]: From the user's point of view, after user submits a scheduled pytorchJob, he will not perceive that we have created a cron job for him, so when the user wants to stop the scheduled training, he can simply delete the pytorchJob to stop it. PropagationDelete is used to configure this function.

File: pkg/job_controller/job.go (hunk 6 above)
Comment (hoaresky) [reply pkg/job_controller/job.go:92]: This causes a problem that every time the job is reconcile it will be queried from the apiserver once

File: pkg/job_controller/job.go (hunk 9 above)
Comment (hoaresky) [reply pkg/job_controller/job.go:391]: Encode newMeta because the cron expression in newMeta has been emptied, so the job template in the newly created cron job has no cron expression. Otherwise the next reconcile job will create a new cron job with cron expressions, resulting in the creation of loops.
RightWay: PytorchJob(cron) --> CronJob(pytorchJob without cron exp) --> PytorchJob
WrongWay: PytorchJob(cron) --> CronJob(pytorchJob with cron exp)  -->  PytorchJob(cron) --> cronJob(pytorchJob with cron exp) --> loop

Comment (jian-he) [reply pkg/job_controller/job.go:391]: my question is why this defer func is needed

File: pkg/job_controller/job.go (hunk 8 above)
Comment (hoaresky) [reply pkg/job_controller/job.go]: removed

File: pkg/job_controller/job.go [hunk 11]
```diff
@@ -324,6 +333,103 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
 	return result, nil
 }
 
+func (jc *JobController) ReconcileCron(meta metav1.Object, runtimeObj runtime.Object, policy *apiv1.RunPolicy) error {
+	cronJob := &appv1.Cron{}
+	// check
+	if policy == nil || policy.CronPolicy == nil{
```
Comment (jian-he) [inline pkg/job_controller/job.go:339]: these two checks can be removed ? because it is already checked in the caller

File: pkg/job_controller/job.go [hunk 12]
```diff
@@ -324,6 +333,103 @@ func (jc *JobController) ReconcileJobs(job interface{}, replicas map[apiv1.Repli
...
+	}
+	return instance, nil
+}
+
+// Extract the submitted job configuration to generate a cron
+// job configuration, and set the parameters that cannot
+// be assigned to empty.
+func resetNonInheritableMetaFields(newMeta metav1.Object) {
+	newMeta.SetName("")
+	newMeta.SetNamespace("")
```
Comment (jian-he) [inline pkg/job_controller/job.go:426]: Is it right to set namespace to empty ? this will cause jobs run in another namespace
And also the name, what will be the new name for subsequenet generated jobs ?

File: pkg/job_controller/job.go (hunk 11 above)
Comment (hoaresky) [reply pkg/job_controller/job.go:339]: I double checked here specially.
if others call this funtion, I'm worried that othes forget to check

File: pkg/job_controller/job.go (hunk 12 above)
Comment (hoaresky) [reply pkg/job_controller/job.go:426]: If these fields are set to empty, CronController will generate a new
https://github.com/kubedl-io/kubedl/blob/master/controllers/apps/cron_controller.go#L327
